회원가입 구현

Hosick opened the pull request:
- 회원 가입 구현
- 닉네임, 이메일 중복 확인
- 휴대폰 인증 기능(SMS 발송은 log로 대체)
- Exception 핸들링

Hosick committed:
[#2]회원가입 구현

-회원가입 구현
-아이디, 이메일 중복검사
-BaseTimeEntity 적용

Hosick committed:
setting branch merge

Hosick committed:
[#2] feat: 회원가입 구현

-DTO를 inner class로 변경
-UserDto Valid 수정
-휴대폰 인증 기능 추가

kihyuk-jeong commented on src/main/java/com/flab/shoeauction/domain/users/Users.java:
```diff
@@ -0,0 +1,38 @@
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Entity
+public class Users extends BaseTimeEntity {
```
여러 명의 회원정보를 저장하고 있는 도메인이 아닌 한 명 한 명의 회원정보가 저장되기 때문에 `Users`보다는 `User`가 맞지 않을까요?

Hosick replied:
앞으로 엔티티가 많이 생길텐데 엔티티의 이름은 단수형이 좋은 것 같습니다. `User`로 변경했습니다.

kihyuk-jeong commented on src/main/java/com/flab/shoeauction/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,57 @@
+                .build();
+        log.error(errorResponse.getMessage(), exception);
+        return new ResponseEntity<>(errorResponse, errorResponse.getStatus());
+    }
```
사용하지 않는 `ExceptionHandler` 같습니다.!

Hosick replied:
비밀번호 체크 기능을 지웠는데 'ExceptionHandler'에는 남아있었네요. 삭제했습니다.

kihyuk-jeong commented on src/main/java/com/flab/shoeauction/web/UserApiController.java:
```diff
@@ -0,0 +1,48 @@
+        if (!usersService.phoneVerification(phone, certificationNumber, session))
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
+        return ResponseEntity.status(HttpStatus.OK).build();
+    }
```
인증과 관련된 보안사항을 `get`방식을 써서 쿼리 스트링으로 받는 것보다는 `post`방식으로 body를 통해 받는 것이 좋지 않을까요??

f-lab-dev replied:
`GET`과 `POST`의 차이는 보안의 관점에서 봐야할까요?

Hosick replied:
세션도 리소스라고 봤을 때 리소스를 생성하는 요청이기때문에 POST가 좋은 것 같습니다.

kihyuk-jeong commented on src/main/java/com/flab/shoeauction/service/users/UsersService.java:
```diff
@@ -0,0 +1,54 @@
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+import javax.servlet.http.HttpSession;
```
`session`은 추후 로그인에도 쓰일 텐데 이런 인증과 관련된 정보들이 저장된 `session`은 따로 클래스를 만들어서 캡슐화하여 관리하는 게 좋을 것 같은데 @Hosick 님은 어떻게 생각하시는지 궁금합니다!

kihyuk-jeong commented on src/main/java/com/flab/shoeauction/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,57 @@
+    @ExceptionHandler(DuplicateEmailException.class)
+    public ResponseEntity<ErrorResponse> duplicateEmailException(DuplicateEmailException exception) {
+        ErrorResponse errorResponse = ErrorResponse.builder()
+                .status(HttpStatus.BAD_REQUEST)
```
400번대 HttpStatus code를 처리하는데 있어서 모두 `BAD_REQUEST`로 처리하는 것과, 중복(리소스 충돌)과 관련있는 `409 Conflict` 과 같은 HttpStatus code로 세분화 하는것에 대해 @Hosick 님과 멘토님의 의견이 궁금합니다.

f-lab-dev replied:
저는 괜찮아보입니다~

Hosick replied:
`409 Conflict`로 수정한다면 클라이언트에서 더 명확하게 알 수 있어서 좋겠네요. 반영하겠습니다~

Hosick committed:
[#2]refactor: 리뷰 반영

-비밀번호 확인 삭제
-인증 post로 변경
-인증 Service 분리

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/util/EncryptionUtils.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.security.MessageDigest;
+
+public class EncryptionUtils {
```
이런 암호화를 하는것도 암호화 방식이 계속 달라질 수 있을 것 같은데 확장성을 위해 인터페이스로 추상화시키는게 어떨까요?

Hosick replied:
인터페이스를 사용해서 결합을 느슨하게 해준다면 암호화 방식이 달라질 때 더욱 쉽게 변경할 수 있을 것 같습니다! 반영하겠습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/exception/ErrorResponse.java:
```diff
@@ -0,0 +1,25 @@
+@Getter
+public class ErrorResponse {
+
+    private HttpStatus status;
```
status는 헤더에 들어가는데 이렇게 하면 body에도 중복으로 들어갈 것 같네요

Hosick replied:
응답에는 에러 메세지만 담는 것으로 수정하였습니다! 그리고 자주 사용하는 응답은 상수로 따로 모아두었습니다

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/CertificationService.java:
```diff
@@ -0,0 +1,42 @@
+@Slf4j
+@RequiredArgsConstructor
+@Service
+public class CertificationService {
```
클래스의 이름은 `CertificationService`인데 내용은 SMS를 이용한 인증이네요. 그럼 `SmsCertificationService` 가 낫지 않을까요?
클래스의 역할이 더 잘 드러나도록 이름을 지어주시면 좋겠습니다~

Hosick replied:
좀더 명확한 이름인 `SmsCertificationService`로 수정하였습니다. 앞으로 네이밍할 때 조금 더 고민해보도록 하겠습니다

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/CertificationService.java:
```diff
@@ -0,0 +1,42 @@
+    public void sendSms(String phone) {
+        String randomNumber = makeRandomNumber();
+        session.setAttribute(phone, randomNumber);
+        log.info("인증번호: " + randomNumber);
```
이 로그는 디버그 용도인것 같은데 `info`로 찍으신 이유가 있으신가요?

Hosick replied:
원래 SMS 발송 API를 이용해서 인증번호를 문자로 보내주는 부분인데, 발송을 대체하는 방식으로 문자대신 log에 띄워주게 되었습니다.  저는 정보성 메세지라고 생각해서 info를 사용하였는데 이러한 부분도 디버그라고 보면 될까요??

f-lab-dev replied:
사용자가 인증할때마다 이 로그가 찍히게 될텐데 이 로그는 우리가 조회해봤을때 서비스를 운영하는데 있어서 도움이 될 수 있는 로그일까요?

f-lab-dev replied:
그리고 로깅하실때 `+` 연산을 하시기보단 문자열 템플릿을 사용해주시는게 좋을 것 같네요~

Hosick replied:
로그 레벨 다시 알아보았고 적절치 않은 로그 사용했던 것 같습니다. 이 부분은 기존 용도에 맞게 SMS 발송으로 변경하였습니다!

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/CertificationService.java:
```diff
@@ -0,0 +1,42 @@
+    // sms 보내기 (log 찍는것으로 대체)
+    public void sendSms(String phone) {
+        String randomNumber = makeRandomNumber();
+        session.setAttribute(phone, randomNumber);
```
SMS의 유효기간은 일반적으로 3분입니다. 이렇게 하면 유효기간이 꽤나 길어질 것 같네요

Hosick replied:
저도 이 부분에서 계속 고민했었는데  세션에 저장된 인증번호에 유효기간을 주려면 세션 자체에 timeout을 주는 방법으로 적용해야할까요?

f-lab-dev replied:
꼭 인증번호는 세션에 저장해야하는걸까요?

Hosick replied:
그럴 필요 없었는데 세션으로 처리해야한다는 생각에 갇혀있던 것 같습니다. 이 부분 조금 더 고민해보겠습니다!

f-lab-dev commented on src/main/java/com/flab/shoeauction/web/UserApiController.java:
```diff
@@ -0,0 +1,52 @@
+
+    private final CertificationService certificationService;
+
+    @GetMapping("/duplication/email")
```
이런 부분은 아래 문서를 읽고 지키도록 노력해주세요

https://docs.microsoft.com/ko-kr/azure/architecture/best-practices/api-design

Hosick replied:
이 부분 조금 더 고민해서 수정하도록 하겠습니다!

Hosick replied:
수정해서 커밋하였습니다!

Hosick committed:
[#2]refactor: 리뷰 반영

-SessionService 생성
-ErrorResponse 삭제
-암호화 수정
-기타 수정

Hosick committed:
[#2]refactor: 리뷰 반영

- SMS 발송 API 적용
- 중복검사 url 수정
- SMS 발송 실패 예외 핸들

f-lab-dev commented on src/main/java/com/flab/shoeauction/util/response/ResponseConstants.java:
```diff
@@ -0,0 +1,19 @@
+
+public class ResponseConstants {
+
+    public static final ResponseEntity<Void> OK_RESPONSE =
```
`ResponseConstants`라는 이름에서 이 변수가 `RESPONSE`라는걸 유추할 수 있습니다.
그러면 뒤에 `_RESPONSE`를 꼭 붙여줘야할까요?

Hosick replied:
이미 클래스명에서 명시해줬으므로 굳이 변수명에 추가로 명시해줄 필요가 없는 것 같습니다. 수정하겠습니다!

f-lab-dev commented on src/main/java/com/flab/shoeauction/web/UserApiController.java:
```diff
@@ -33,20 +34,19 @@
     }
 
     @PostMapping("/certification/sms")
     public ResponseEntity<Void> sendSms(@RequestBody Map<String, String> map) {
```
여기선 명시적인 클래스가 아니라 `Map`을 쓰신 이유가 무엇이실까요? 그리고 이런 곳에서 `Map`을 쓰면 유지보수성 관점에서 어떤 문제가 발생할까요?

Hosick replied:
앗 저 부분은 `@RequestBody`로 String 하나만 받았을 때 문자열에 중괄호가 붙는 문제가 생겨서 잠깐 Map으로 변경했다가 다시 Dto로 받는 것으로 변경해서 커밋했었습니다. 

 만약 Map으로 받게된다면 직접 key 값을 다뤄야 하기 때문에 값들을 사용하려할 때(특히 다른 사람이) 무슨 값이 들어있는지 파악하기가 어렵다는 문제가 있을 것 같습니다. 정형화된 DTO를 사용하는 방법이 훨씬 낫다고 생각됩니다.

Hosick committed:
[#2]refactor: 회원가입 리뷰 반영 수정

-ResponseConstants 변수명 수정
-SMS 인증 제한시간 구현

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/SessionService.java:
```diff
@@ -1,27 +1,27 @@
 
-    public void saveSmsCertificationNumber(String number) {
-        session.setAttribute(SMS_CERTIFICATION_NUBER, number);
+    public void saveSmsCertificationInfo(SentSmsCertificationInfo sentSmsCertificationInfo) {
```
인증정보를 세션에 두는건 어떻게 바꾸실 생각이실까요?

Hosick replied:
인증정보를 DB에 저장하는 방법 또한 생각해 보았습니다. 하지만 기존에 사용하는 디스크 기반 RDB에 저장하게 된다면 단기성 정보인 인증 정보가 쌓이기 때문에 오히려 세션에 저장하는 방식보다 좋지 않을거라는 생각이 들었습니다.
 그렇다면 Redis와 같은 인 메모리 DB에 담는 방법을 사용한다면 적절할까요??

f-lab-dev replied:
네 그게 나을 것 같네요~

f-lab-dev replied:
이건 기혁님께 남긴것과 마찬가지인데 그러면 이 브랜치에서는 SMS인증과 회원가입을 동시에 개발하고 있는거네요.
그러면 이 브랜치는 한가지 이슈에 대해서 개발하는 브랜치가 아니게 되는거고요~

Hosick replied:
맞는 것 같습니다! 
이슈 분리하고 앞으로는 이슈 최대한 명확하게 나누겠습니다!

Hosick committed:
[#2]chore: SMS 인증 이슈 분리

-회원가입 이슈에서 SMS인증 이슈 분리

Hosick committed:
[#8] 회원가입 시 SMS 인증 구현

-redis 기본 설정
-인증번호 발송 시 redis에 휴대폰 번호와 인증번호를 key:value 형식으로 저장
-메모리에 저장된 인증번호 값에 제한시간  부여

Hosick committed:
merge settings branch

Hosick committed:
[#8] 리뷰 반영

-인증번호 prefix 변수명 수정
-redis set, expire 명령 하나로 통합

Hosick committed:
Merge branch 'feature/8/hs' into feature/2/hs
